### PR TITLE
Fix: propagate npm ci errors during deploy

### PR DESCRIPTION
**Pull request by grncdr**
The `deploy` command currently shells out to `npm ci --ignore-scripts` in each lambdas directory, but it doesn't check whether npm succeeded. This leads to a really nasty behaviour where lambdas can be deployed with missing dependencies.

I can reliably reproduce this locally by running `npx deploy` in a project with about a dozen lambdas. Before this change, some of my lambdas would be broken after every deploy. After this change I see errors like this:

```
Error: npm ci in <snip: /path/to/my/lambda> exited with code 1
npm WARN prepare removing existing node_modules/ before installation
fs.js:667
  return binding.open(pathModule.toNamespacedPath(path),
                 ^

Error: ENFILE: file table overflow, open '/usr/local/lib/node_modules/npm/node_modules/minipass/index.js'
   <snip: stack trace in npm>
```

This is a bit ugly, but a hell of a lot better than deploying a broken lambda 😄 

### checklist

- [x] Forked the repository and created your branch from master
- [x] Make sure the tests pass! `npm it` in the repository root
- [ ] If you've fixed a bug or added code **more** tests are appreciated

I don't know how to start testing this... I guess we could set up a lambda with an invalid package.json or something else that would guarantee npm to fail?

### CODE REVIEW

**Comment by grncdr** [comment]
Generally speaking, it seems to me that running `npm ci` that many times in parallel is unlikely to work well (my fans go crazy every time I deploy). What about limiting it to 4 at a time or so?

**Comment by brianleroux** [comment]
yea this is weird! we deploy begin.com with about 60 lambdas without issue but our dep tree in each is pretty small. how many deps do your lambdas have? also: what version of npm are you using? (we def see the fans issue and I'm totally into the idea of throttling npm / we could prototype that with an env var pretty easily)

**Comment by grncdr** [comment]
My dep tree in these was also pretty small, pretty much only using arc functions, has id's and tiny-json-http 

I actually implemented the concurrency limit on my work by just swapping parallel for parallel-limit and that solved it for me

**Comment by ryanblock** [comment]
A couple thoughts: running `npm ci` on many (ie dozens+) of functions simultaneously is indeed a pretty intense operation, but also one that is generally pretty deterministic thanks to `package-lock.json` manifests (despite many misgivings and frustrations there).

In our experience having shipped numerous fairly large Architect apps – and deployed them many, many times over the last two years – we've occasionally seen npm problems during deploy/hydration in two scenarios: npm connectivity issues (of course), and missing (or corrupt) `package-lock.json` files.

In these cases, Architect is known to sometimes swallow errors and can produce bad deploys. See also:
- [During deploys, bubble npm errors and abort](https://github.com/arc-repos/architect/issues/141)
- [Areas for improvement in deployment workflow](https://github.com/arc-repos/architect/issues/151)

With that in mind, I was able to verify this patch by removing some functions' `package-lock.json` files, as well as by munging them (i.e. modifying their sha values, etc.). This patch caught both scenarios in both full and single-Lambda deployments. Without this patch in place, 4.4.9 can still potentially deploy a function that would be rendered broken immediately upon deployment without its modules.

tl;dr I'm all for this: we really doing a better job of bubbling npm errors, this seems like a good start. 

@grncdr I'd like to merge, possible to trouble you to look at a functional equivalent for `hydrate`? This is often the first step for onboarding a new Architect project (or when collaborating with others and a new function is added in source control). See: https://github.com/arc-repos/architect/blob/master/src/hydrate/index.js

**Comment by grncdr** [comment]
>  I'd like to merge, possible to trouble you to look at a functional equivalent for `hydrate`?

I did that, then noticed that not only did I miss an undefined var, but also the other usage of npm in the same file. All the repetition was getting to me a bit so...

Now I've added a [utility function](https://github.com/arc-repos/architect/pull/231/commits/8d07e2825ca17d74e696cd0681a346d95e6d7b6f#diff-6b13064cc32515820f4bd32ff705516eR5) in `src/util/npm.js` which does the stderr aggregation and error forwarding, and replaced all the places that were spawning their own subprocesses with calls to that.

After that, I just went ahead and added a [concurrency limiter](https://github.com/arc-repos/architect/pull/231/commits/6bb02c316ee97cf3d504d19a833d36fb7cdfeb25), which should address the overall problem and make spawning a lot of npm processes more reliable and more friendly on system resources.

I don't have so many lambda in my project yet, so if you experiment with the `ARC_MAX_NPMS` env variable I'd be curious to hear how that impacts performance for e.g. `npx hydrate`.

**Comment by ryanblock** [comment]
Sorry for the hassle – I went to validate your fix and we have a bug or tw  that is preventing hydrate from running. I believe I fixed since you cut this branch, so if you rebase I should be able to pick up and continue testing.

That said, on the deploy side my initial test appeared to regress; I followed the same validation step as I outlined before (munged a `package-lock.json` file in a function, then deployed) but it didn't catch the error, and instead deployed missing a dep. Sorry I don't have more to give you there at the moment, haven't yet had a chance to look deeper into your changes to figure out where the regression is.

**Comment by ryanblock** [comment]
Had a chance to chat about this PR with @brianleroux this morning, and we agreed that directionally we'd like to ensure that the rest of the deploy completes even if a single function's npm operations error out. Will stand by for you to clear the other stuff, and happy to help out.

Related, would you a contributor invite? Loving what you're doing here!

**Comment by ryanblock** [comment]
Just a heads up: a few changes (that should be easy to work around / integrate) from today related to both `deploy` and `hydrate`:
- https://github.com/arc-repos/architect/commit/0673b26693bbcde2de5e1201ec4cb28a73a94060
- https://github.com/arc-repos/architect/commit/c38ad60d90d43c28777ea2ff338617bf6298d070

**Comment by grncdr** [comment]
> That said, on the deploy side my initial test appeared to regress; I followed the same validation step as I outlined before (munged a package-lock.json file in a function, then deployed) but it didn't catch the error, and instead deployed missing a dep.

Oh no! I'll look into what's happening there.

> Had a chance to chat about this PR with @brianleroux this morning, and we agreed that directionally we'd like to ensure that the rest of the deploy completes even if a single function's npm operations error out.

I don't _think_ this PR should affect that behaviour, which afaict is handled in [`lambda-all/01-prep.js`](https://github.com/arc-repos/architect/blob/c560ea6ba82d3078dd7361fdc55525fb5db45c42/src/deploy/lambda-all/01-prep.js#L54).

> Related, would you like a contributor invite? Loving what you're doing here!

Sure! Though I'll prefer to keep working in my own fork for now. I've rebased my branch and will incorporate the changes for https://github.com/arc-repos/architect/commit/0673b26693bbcde2de5e1201ec4cb28a73a94060 today.

**Comment by ryanblock** [comment]
Sweet. Contributor invite sent, welcome! 🌈

Digging into this PR (and the stuff that shipped today in Arc 4.4.10), I think we're very soon going to need to refactor into a generic, vendor-independent dependency hydrator. (That's not to imply that the scope of what you're doing here should expand, just want to give context / roadmap.)

Right now `deploy` has its own hydration flow separate from `hydrate`, and this is only going to get more unwieldy when we ship Python support (with other runtimes to follow).

I don't think this changes what you're doing here much or at all right now, but I think we'll probably want to consolidate all dependency hydration logic into `src/hydrate`, with vendors like `src/hydrate/npm` or `src/hydrate/providers/npm`.

Thoughts?

**Comment by grncdr** [comment]
https://github.com/arc-repos/architect/pull/230#issuecomment-449338281

**Comment by ryanblock** [comment]
Hey @grncdr, just wanted to say thanks again for your help on this! Your stuff is in Architect `4.5.0`, released today. I am soooo much happier with our package manager (and related error) handling now.

P.S. -The official env var for the npm queue is `ARC_MAX_NPM` fyi (defaults to `10`).

🌈🌈🌈

**Comment by grncdr** [comment]
Thanks for the heads up @ryanblock 😄 I just got back from vacation so I'll look into the new architect versions sometime soon.
